### PR TITLE
[WIP]: Integrate Ultralytics Support

### PR DESCRIPTION
**Pull request by deependujha**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -202,7 +202,7 @@ ld.map(
 ## Features for optimizing and streaming datasets for model training
 
 <details>
-  <summary> ✅ Stream large cloud datasets</summary>
+  <summary> ✅ ge cloud datasets</summary>
```
Comment (deependujha) [inline README.md]:
```suggestion
  <summary> ✅ Stream large cloud datasets</summary>
```

File: README.md [hunk 2]
```diff
@@ -1514,6 +1514,93 @@ if __name__ == "__main__":
 
 </details>
 
+<details>
+  <summary>✅ Stream Large Datasets to Ultralytics Models with LitData</summary>
```
Comment (tchaton) [inline README.md:1560]: Let's create an integration section with Ultralytics. And let's make it more clear what are the benefits which is training in the cloud.

File: src/litdata/streaming/dataset.py [hunk 3]
```diff
@@ -441,7 +450,23 @@ def __getitem__(self, index: Union[ChunkedIndex, int, slice]) -> Any:
...
+                        param.kind == inspect.Parameter.VAR_KEYWORD for param in signature.parameters.values()
+                    )
+                    item = transform_fn(item, **self.transform_kwargs) if accepts_kwargs else transform_fn(item)
+            else:
+                # check if transform function accepts kwargs
+                signature = inspect.signature(self.transform)
+                accepts_kwargs = any(
+                    param.kind == inspect.Parameter.VAR_KEYWORD for param in signature.parameters.values()
+                )
+                item = self.transform(item, **self.transform_kwargs) if accepts_kwargs else self.transform(item)
```
[the remark this answers is not in the document]
Comment (deependujha) [reply src/litdata/streaming/dataset.py]: `transform` method can also be implemented in a derived class.

File: src/litdata/integrations/ultralytics/patch.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+        from ultralytics.utils import ops
+
+        # TODO: optimize this
+        for i, pred in enumerate(preds):
+            pred["batch_idx"] = torch.ones_like(pred["conf"]) * i  # add batch index to predictions
+        keys = preds[0].keys()
+        max_det = max_det or self.args.max_det
+        batched_preds = {k: torch.cat([x[k][:max_det] for x in preds], dim=0) for k in keys}
+        # TODO: fix this
+        batched_preds["bboxes"][:, :4] = ops.xyxy2xywh(batched_preds["bboxes"][:, :4])  # convert to xywh format
```
[the remark this answers is not in the document]
Comment (deependujha) [reply src/litdata/integrations/ultralytics/patch.py:245]: taken directly from ultralytics code, so don't wanna mess up with it.

File: README.md (hunk 2 above)
Comment (deependujha) [reply README.md:1560]: done. Also added:

Instead of downloading entire datasets (which can be hundreds of GBs), you can now **stream data on-the-fly from S3, GCS, HTTP(S), or even local disk** — making it ideal for training in the cloud with limited storage and more efficient utilization of resources.

File: src/litdata/integrations/ultralytics/optimize.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    if img_ext in [".jpg", ".jpeg", ".png"]:
+        # Reduce quality to 90%
+        encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 90]
+        success, encoded = cv2.imencode(".jpg", img, encode_param)
+        if not success:
+            raise ValueError(f"JPEG encoding failed for: {img_path}")
+
+        # Decode it back to a numpy array (OpenCV default format)
+        img = cv2.imdecode(encoded, cv2.IMREAD_COLOR)
+
```
Comment (bhimrazy) [inline src/litdata/integrations/ultralytics/optimize.py:50]: I think we might not need the decode step and can store the images as JPEGs instead—maybe using PIL or `torchvision`.

Note from @tchaton . 
```
torchvision jpeg_encode and decode is the fastest
```

Comment (bhimrazy) [reply src/litdata/integrations/ultralytics/optimize.py:50]: maybe, we can also provide an option for image quality, with 90 as the default.

File: src/litdata/integrations/ultralytics/optimize.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        return None  # skip this sample
+
+    return {
+        "img": img,
+        "label": label,
+    }
+
+
+def optimize_ultralytics_dataset(
+    yaml_path: str,
```
Comment (bhimrazy) [inline src/litdata/integrations/ultralytics/optimize.py:67]: a query about the YAML file — where can we find it? Is it something that's already available, or does the user need to create their own?

File: src/litdata/integrations/ultralytics/optimize.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    # parse the YAML file & make sure data exists, else download it
+    dataset_config = check_det_dataset(yaml_path)
+
+    output_dir = _resolve_dir(output_dir)
+
+    mode_to_dir = {}
+
+    for mode in ("train", "val", "test"):
+        if dataset_config[mode] is None:
+            continue
```
Comment (bhimrazy) [inline src/litdata/integrations/ultralytics/optimize.py:114]: Here, could there be a case where the test set is absent?

File: src/litdata/processing/data_processor.py [hunk 8]
```diff
@@ -1089,6 +1089,7 @@ def __init__(
         start_method: Optional[str] = None,
         storage_options: Dict[str, Any] = {},
         keep_data_ordered: bool = True,
+        verbose: bool = True,
```
Comment (bhimrazy) [inline src/litdata/processing/data_processor.py]: I think it might be better to keep this part in a separate PR to get it landed, without making any changes to the LitData internals in this PR, wdyt ? @deependujha

File: src/litdata/streaming/dataset.py [hunk 9]
```diff
@@ -89,7 +92,10 @@ def __init__(
                 If `index_path` is a directory, the function will look for `index.json` within it.
                 If `index_path` is a full file path, it will use that directly.
             force_override_state_dict: Boolean flag for allowing local arguments to override a loaded state dict.
-            transform: Optional transformation function to apply to each item in the dataset.
+            transform: Optional transformation function or list of functions to apply to each item in the dataset.
+            transform_kwargs: Keyword arguments for the transformation function.
+            args: Additional positional arguments.
+            kwargs: Additional keyword arguments.
```
Comment (bhimrazy) [inline src/litdata/streaming/dataset.py]: It would be nice to also separate this change into a new quick PR, including details, examples, and motivation for the update.

**Review by bhimrazy — COMMENTED** [review]
Added a few comments and queries. Looking great so far!

File: src/litdata/integrations/ultralytics/optimize.py (hunk 6 above)
Comment (deependujha) [reply src/litdata/integrations/ultralytics/optimize.py:67]: You can find yaml files for existing datasets in [ultralytics repo](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/cfg/datasets)

And, you can also write for your own custom dataset: https://docs.ultralytics.com/datasets/#contribute-new-datasets

File: src/litdata/integrations/ultralytics/optimize.py (hunk 7 above)
Comment (deependujha) [reply src/litdata/integrations/ultralytics/optimize.py:114]: Possibly, yes, especially with custom datasets. Some users may choose to include only a validation set and skip the test set entirely. It depends on the training workflow or evaluation preference.

File: src/litdata/processing/data_processor.py (hunk 8 above)
Comment (deependujha) [reply src/litdata/processing/data_processor.py]: sure. Makes perfect sense.

File: src/litdata/integrations/ultralytics/optimize.py (hunk 5 above)
Comment (deependujha) [reply src/litdata/integrations/ultralytics/optimize.py:50]: let's do this in next pr.

File: src/litdata/integrations/ultralytics/optimize.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+    # Read image using OpenCV
+    img = cv2.imread(img_path, cv2.IMREAD_COLOR)
+    if img is None:
+        raise ValueError(f"Failed to read image: {img_path}")
+
+    # JPEG re-encode if image is jpeg or png
+    if img_ext in [".jpg", ".jpeg", ".png"]:
+        # Reduce quality to 90%
+        encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 90]
```
Comment (lantiga) [inline src/litdata/integrations/ultralytics/optimize.py]: I would expose image quality as a configurable argument.

Reducing image quality by default might be ok for generic tasks on natural images, but for other applications it might induce a bias in the model and a potential mismatch when deploying the trained model for inference.

File: src/litdata/integrations/ultralytics/patch.py [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+
+import numpy as np
+import torch
+
+from litdata.constants import _ULTRALYTICS_AVAILABLE
+from litdata.streaming.dataloader import StreamingDataLoader
+from litdata.streaming.dataset import StreamingDataset
+
+
+def patch_ultralytics() -> None:
```
Comment (lantiga) [inline src/litdata/integrations/ultralytics/patch.py:26]: There's a bit of a licensing pickle here, since Ultralytics has a custom license and we are taking code from it here in order to patch it. I don't think this is a material problem per se, since we are enhancing the use of the Ultralytics library and not copying functionality. I would still point it out in a comment in the file, and try to minimize the amount of Ultralytics code we copy here.

In addition, I would add a few regression tests that will break if the API in Ultralytics changes, because we are making quite a few assumptions on the way Ultralytics is laid out today.

**Review by lantiga — COMMENTED** [review]
Great work, added a couple of comments

**Comment by deependujha** [comment]
Closing this PR as the decision was made to move this integration and similar future ones to a `private repository`, both to **`prevent potential licensing conflicts`** and to **`manage customer-specific distributions`**.

I've recreated the PR in the private repo as requested.

Thanks for the reviews.
